### ISSUE TITLE
The forward accuracy cannot be aligned when using  DTensor and ColwiseParallel, RowwiseParallel for TP training on the Llama2

### ISSUE THREAD

**Issue by wangbluo** [pytorch/pytorch#120158]
### 🐛 Describe the bug

Hi, I found that the forward accuracy cannot be aligned completely when using your 'DTensor' and 'ColwiseParallel', 'RowwiseParallel', and 'Replicated' api for TP training on the Llama2. 

I referred your example: https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_tp_examples.py#L175C9-L262C1. I parallelize the embedding submodules of Llama2 by 'Replicate()', parallelize the attention and feed forward by 'ColwiseParallel' and 'RowwiseParallel', and Replicate the model.lm_head for the output, and manually adjust the number of heads and the hidden size.

The following is my code:

```
def get_tensor_sharded_model(model, use_ddp):

    device = "cuda" if torch.cuda.is_available() else "cpu" 
    device_mesh = DeviceMesh(device, torch.arange(0, NUM_DEVICES))
    # Parallelize the embedding submodules. 
    parallelize_module(model.model.embed_tokens, device_mesh, ColwiseParallel(output_layouts=Replicate()))

    # Parallelize the attention and feed forward submodules.
    for layer in model.model.layers:
        layer_parallelize_plan = {} 
        layer_parallelize_plan["self_attn.q_proj"] = ColwiseParallel()
        layer_parallelize_plan["self_attn.k_proj"] = ColwiseParallel()
        layer_parallelize_plan["self_attn.v_proj"] = ColwiseParallel()
        layer_parallelize_plan["self_attn.o_proj"] = RowwiseParallel()

        layer_parallelize_plan["mlp.gate_proj"] = ColwiseParallel()
        layer_parallelize_plan["mlp.up_proj"] = ColwiseParallel()
        layer_parallelize_plan["mlp.down_proj"] = RowwiseParallel()

        parallelize_module(layer, device_mesh, layer_parallelize_plan)

    # Parallelize the output submodule. 
    output_parallelize_plan = RowwiseParallel(input_layouts=Replicate())
    parallelize_module(model.lm_head, device_mesh, output_parallelize_plan)

    # Manually adjust the number of heads after sharding the self attention modules.
    for layer in model.model.layers:
        assert model.model.config.num_attention_heads % dist.get_world_size() == 0
        # For llama2 models, your should adjust the number of heads separately.
        layer.self_attn.num_heads = model.model.config.num_attention_heads // dist.get_world_size()
        layer.self_attn.num_key_value_heads = model.model.config.num_key_value_heads // dist.get_world_size()
        layer.self_attn.hidden_size = model.model.config.hidden_size // dist.get_world_size()
```

And here is my loss comparison  between model and model_tp:
![img_v3_0284_103fb69e-539e-427f-8d3f-29e7981c5f7g](https://github.com/pytorch/pytorch/assets/32676639/7fb6593a-6f61-4d88-8ffe-78bf742716c6)
and my comparison code, just like you used in your example:
![img_v3_0284_72d46648-041d-474f-9004-909adbe6764g](https://github.com/pytorch/pytorch/assets/32676639/18510bc6-8466-4004-a9aa-556cb461d0a4)

It said I have 29% elements mismatched in my first step loss comparison!

I feel very confused and have invested the whole process, like whether the input is consistent across each rank(it is), and whether the dropout in the llama2 model could introduce randomness (Actually llama2 does not have dropout). Now I really don't know why. Could anybody help me to check this? Should I add some all_reduce for the loss or something?

### Versions

Package                 Version            Editable project location
----------------------- ------------------ -------------------------
absl-py                 2.1.0
accelerate              0.26.1
aiohttp                 3.9.3
aiosignal               1.3.1
astunparse              1.6.3
attrs                   23.2.0
blessed                 1.20.0
cachetools              5.3.2
certifi                 2024.2.2
charset-normalizer      3.3.2
datasets                2.16.1
dill                    0.3.7
expecttest              0.2.1
filelock                3.13.1
frozenlist              1.4.1
fsspec                  2023.10.0
google-auth             2.26.2
google-auth-oauthlib    1.2.0
gpustat                 1.1.1
grpcio                  1.60.0
huggingface-hub         0.20.3
hypothesis              6.97.4
idna                    3.6
Jinja2                  3.1.3
lark                    1.1.9
Markdown                3.5.2
MarkupSafe              2.1.5
mpmath                  1.3.0
multidict               6.0.5
multiprocess            0.70.15
networkx                3.2.1
numpy                   1.26.3
nvidia-ml-py            12.535.133
oauthlib                3.2.2
optree                  0.10.0
packaging               23.2
pandas                  2.2.0
pip                     23.3.1
prettytable             3.9.0
protobuf                4.23.4
psutil                  5.9.8
pyarrow                 15.0.0
pyarrow-hotfix          0.6
pyasn1                  0.5.1
pyasn1-modules          0.3.0
python-dateutil         2.8.2
pytorch-triton          3.0.0+901819d2b6
pytz                    2024.1
PyYAML                  6.0.1
regex                   2023.12.25
requests                2.31.0
requests-oauthlib       1.3.1
rsa                     4.9
safetensors             0.4.2
sentencepiece           0.1.99
setuptools              68.2.2
six                     1.16.0
sortedcontainers        2.4.0
sympy                   1.12
tensorboard             2.15.1
tensorboard-data-server 0.7.2
tokenizers              0.15.1
torch                   2.3.0a0+gitfcf22a8 /home/wangbinluo/pytorch
tqdm                    4.66.1
transformers            4.37.2
types-dataclasses       0.6.6
typing_extensions       4.9.0
tzdata                  2023.4
urllib3                 2.2.0
wcwidth                 0.2.13
Werkzeug                3.0.1
wheel                   0.41.2
xxhash                  3.4.1
yarl                    1.9.4

**Comment by wangbluo** [comment]
@wanchaol Hello, would you mind assisting me with checking the accuracy on TP training with llama2? I really don't know whether it's wrong with my code or something. I would be grateful if you could help me.

**Comment by wangbluo** [comment]
By the way, the weight of model and model_tp before forward can be alighted. So at least we can narrow down the scope of the problem to the first forward pass of the model. I think the problem probably be the all_reduce.

![img_v3_0286_06138c6e-460a-4ba8-85cf-bf7ad6c4cb7g](https://github.com/pytorch/pytorch/assets/32676639/71739ed0-475d-497c-aeed-22d4a7cc15b3)
![img_v3_0286_e3f1f3dc-a2b3-441f-9226-fb061ff29e8g](https://github.com/pytorch/pytorch/assets/32676639/47f9edbd-2a41-4c83-b051-e9932d238e9f)

**Comment by wangbluo** [comment]
Hi, dear developers, I have narrow down it's after the MLP of Llama2 model, the accuracy cannot be aligned totally. There is a SiLU activation in MLP, which is different from your example. I'm trying to solve it and hope it could be aligned later.

**Comment by wangbluo** [comment]
Hello, there is a correction. 
I found that during the llama2 model forward pass, the precision can be aligned after **the first LlamaDecoderLayer**. However, after passing through **24 layers** of LlamaDecoderLayer class, this precision gap increases to **11%.** This is beyond my understanding. 
You can try looping the transformer layer in your example multiple times instead of just one layer to see if the same precision issue occurs.

**Comment by wangbluo** [comment]
And I used torch.float32 to do the training.

**Comment by wconstab** [comment]
@wanchaol @tianyu-l can one of you take a look?

**Comment by wangbluo** [comment]
Hello, I think I have solved the forward accuracy. It's the reason that you use the AsyncTensor to do the communication for TP. **The AsyncTensor is still dangerous now,** and can lead some errors in the compiler, as @wconstab mentioned in the comment https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L119

I use native c10d api to replace them and now the forward accuracy can be aligned, the `testcase.assertEqual(output[0], output_tp[0]) ` code can be passed now.

![image](https://github.com/pytorch/pytorch/assets/32676639/a0b7a2b8-a31d-4ec6-bdcf-bfcc78b73b3a)

But the backward of model and model_tp still cannot be aligned now, it said there are  **Mismatched elements: 7 / 4194304 (0.0%)** of attention weights between model and model_tp. :(
The corresponding elements of the model and model_tp's weights looks like very close but not the same. Maybe I should check the backward of attention.
I am still working on finding a solution. I would greatly appreciate any suggestions you may have.
![image](https://github.com/pytorch/pytorch/assets/32676639/7402fcf6-9415-4a60-a15e-a3fc6dd98271)

**Comment by wangbluo** [comment]
The loss curve still cannot be converged. 😢 Would anyone be so kind as to lend a hand? I would be extremely grateful. Thank you!

**Comment by wconstab** [comment]
My comment about tracing asynctensor just means the compiler starts with an unoptimized graph from timing/overlap perspective and must do its own reordering to improve performance. I don't know of an accuracy problem with FC/compile. But if you're reporting one we'll take a look. We might need some help dumping inductor IR graphs. Cc @yifuwang

**Comment by wangbluo** [comment]
> Hey @wangbluo, just a heads up that we landed a functional collective overhaul a few days ago (#120370). Can you check if your local PyTorch build contains the change? If not, I'd suggest updating your PyTorch build and re-running your experiments with the change. It's not guaranteed, but I think there's a good chance it may address your issue.
> 
> If the issue persists, please provide a minified repro. I'd be happy to look into it.

Thank you for your crucial reply! It really works for the forward accuracy but still can't make the backward accuracy aligned after I update my local PyTorch to the latest including #120370.

Sure! You can find the start.sh script at this link: [start.sh](https://github.com/wangbluo/Finetune_llama2/blob/main/start.sh). Don't worry about the model and data paths as they are all downloaded from Hugging Face.

The tensor parallel part can be found in [tensor_parallel.py](https://github.com/wangbluo/Finetune_llama2/blob/main/tensor_parallel.py), and the tensor parallel unit test part is available in [finetune_llama2.py](https://github.com/wangbluo/Finetune_llama2/blob/main/finetune_llama2.py#L192C5-L218C48). Both are referenced from the tp_examples: [test_tp_examples.py](https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_tp_examples.py#L175C9-L262C1).
